crates/concrete_check/src/linearity_check.rs

```diff
@@ -0,0 +1,1086 @@
+        let vars = state_tbl.vars.clone();
+        println!("Check_expr vars {:?}", vars);
+        for (name, _info) in vars.iter() {
+            //self.check_var_in_expr(depth, &name, &info.ty, expr)?;
+            state_tbl = self
+                .check_var_in_expr(state_tbl, depth, name, expr, context)
+                .unwrap();
+        }
+        Ok(state_tbl)
+    }
+
+    fn count_in_statements(&self, name: &str, statements: &[Statement]) -> Appearances {
+        statements
+            .iter()
+            .map(|stmt| self.count_in_statement(name, stmt))
+            .fold(Appearances::zero(), |acc, x| acc.merge(&x))
+    }
+
+    fn count_in_statement(&self, name: &str, statement: &Statement) -> Appearances {
+        match statement {
+            Statement::Let(binding) => {
+                // Handle let bindings, possibly involving pattern matching
+                self.count_in_expression(name, &binding.value)
+            }
+            Statement::If(if_stmt) => {
+                // Process all components of an if expression
+                let cond_apps = self.count_in_expression(name, &if_stmt.value);
+                let then_apps = self.count_in_statements(name, &if_stmt.contents);
+                let else_apps;
+                let else_statements = &if_stmt.r#else;
+                if let Some(else_statements) = else_statements {
+                    else_apps = self.count_in_statements(name, else_statements);
+                } else {
+                    else_apps = Appearances::zero();
+                }
+                cond_apps.merge(&then_apps).merge(&else_apps)
+            }
+            Statement::While(while_expr) => {
+                let cond = &while_expr.value;
+                let block = &while_expr.contents;
+                // Handle while loops
+                self.count_in_expression(name, cond)
+                    .merge(&self.count_in_statements(name, block))
+            }
+            Statement::For(for_expr) => {
+                // Handle for loops
+                //init, cond, post, block
+                let init = &for_expr.init;
+                let cond = &for_expr.condition;
+                let post = &for_expr.post;
+                let block = &for_expr.contents;
+                let mut apps = Appearances::zero();
+                if let Some(init) = init {
+                    if let Some(cond) = cond {
+                        if let Some(post) = post {
+                            apps = self
+                                .count_in_let_statements(name, init)
+                                .merge(&self.count_in_expression(name, cond))
+                                .merge(&self.count_in_assign_statement(name, post))
+                                .merge(&self.count_in_statements(name, block))
+                        }
+                    }
+                }
+                apps
+            }
+            /* Alucination of GPT
```

Comment:
this comment does not belong here. 
We should check the original algorithm as the source of truth

Reply:
GPT4 infered Statement::Block exists, but it doesn't exists. Remved the comment